### PR TITLE
Added Binary Exponentiation in math directory

### PR DESCRIPTION
**Pull request by mann2108**
#### Description of Change
Binary Exponentiation is the method to calculate the a^n (where '^' denotes the power operator). This will find a^n in just O(log(n)) complexity instead O(n) multiplication. The idea behind this algorithm is it will convert the exponent into binary which came up with exactly log(n)+1 digits in base 2.

Example: -
2^10 = 2^(1010) = 2^8 * 2^2

2^1 = 2
2^2 = (2^1)^2 = 2^2 = 4
2^4 = (2^2)^2 = 4^2 = 16
2^8 = (2^4)^2 = 16^2 = 256

Hence to calculating 2^10 we only need to multiply 2^8 and 2^2 skipping 2^1 and 2^4.

#### Checklist

- [X] Added description of change
- [X] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
- [X] Added tests and example, test must pass
- [X] Relevant documentation/comments is changed or added
- [X] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
- [X] Search previous suggestions before making a new one, as yours may be a duplicate.
- [X] Sort by alphabetical order
- [X] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by mann2108** [comment]
@cclauss I request you to review this PR :)

File: math/binary_exponent.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/// C++ Program to find Binary Exponent recursively.
+
+#include<iostream>
+/*
+ * Calculating a^b in O(log(b)) by converting b in binary no.
```
Comment (poyea) [inline math/binary_exponent.cpp:5]:
```suggestion
 * Calculate a^b in O(log(b)) by converting b to a binary number
```

File: math/binary_exponent.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * Binary exponentiation (also known as exponentiation by squaring)
+ * is a trick which allows to calculate an using only O(logn) multiplications
```
Comment (poyea) [inline math/binary_exponent.cpp:7]: Please review this sentence.

File: math/binary_exponent.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    if (b%2) {
+        return res*res*a;
+    } else {
+        return res*res;
+    }
+}
+
+int main() {
+    int a, b;
+    /// Give two nos as a^b (where '^' denotes power exponent operation)
```
Comment (poyea) [inline math/binary_exponent.cpp:26]:
```suggestion
    /// Give two numbers a, b
```

File: math/binary_exponent.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    std::cin >> a >> b;
+    /// Result of a^b
```
Comment (poyea) [inline math/binary_exponent.cpp:28]:
```suggestion
    /// Result of a^b  (where '^' denotes exponentiation)
```

**Comment by poyea** [comment]
> #### Description of Change
> Binary Exponentiation is the method to calculate the a^n (where '^' denotes the power operator). This will find a^n in just O(log(n)) complexity instead O(n) multiplication. The idea behind this algorithm is it will convert the exponent into binary which came up with exactly log(n)+1 digits in base 2.
> 
> Example: -
> 2^10 = 2^(1010) = 2^8 * 2^2
> 
> 2^1 = 2
> 2^2 = (2^1)^2 = 2^2 = 4
> 2^4 = (2^2)^2 = 4^2 = 16
> 2^8 = (2^4)^2 = 16^2 = 256
> 
> Hence to calculating 2^10 we only need to multiply 2^8 and 2^2 skipping 2^1 and 2^4.
> 
> #### Checklist
> * [x]  Added description of change
> * [x]  Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
> * [x]  Added tests and example, test must pass
> * [x]  Relevant documentation/comments is changed or added
> * [x]  PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
> * [x]  Search previous suggestions before making a new one, as yours may be a duplicate.
> * [x]  Sort by alphabetical order
> * [x]  I acknowledge that all my contributions will be made under the project's license.

The example here can be inserted into the code as well.

**Comment by mann2108** [comment]
@poyea Updated with requested changes.

File: math/binary_exponent.cpp (hunk 2 above)
Comment (mann2108) [reply math/binary_exponent.cpp]: ok.

File: math/binary_exponent.cpp [hunk 5]
```diff
@@ -1,14 +1,21 @@
...
+ * 2^1 = 2
+ * 2^2 = (2^1)^2 = 2^2 = 4
+ * 2^4 = (2^2)^2 = 4^2 = 16
+ * 2^8 = (2^4)^2 = 16^2 = 256
+ * Hence to calculate 2^10 we only need to multiply 2^8 and 2^2 skipping 2^1 and 2^4.
 */
 
-/// Function to calculate exponent in O(log(n)) using binary exponent.
+/// Recursive function to calculate exponent in O(log(n)) using binary exponent.
 int binExpo(int a, int b) {
```
Comment (cclauss) [inline math/binary_exponent.cpp:19]: Why make your reader guess??  Why not `binary_exponent_recursive()` and `binary_exponent_iterative()`?

Comment (cclauss) [reply math/binary_exponent.cpp:19]: Too late.

Comment (poyea) [reply math/binary_exponent.cpp:19]: It's clear enough up to the scope of this file (see `main()`). When one uses it else where, this is of course the best version he can get.

Comment (mann2108) [reply math/binary_exponent.cpp:19]: I agreed with @cclauss. I think its more better to give proper justifications in function name itself. what you say? @cclauss @poyea 
So should I open a new PR or else we are going with this version?

**Comment by mann2108** [comment]
> That's good! Thank you for your pull request!🤩

Thank you. I will come soon with a new algorithm. :)

**Pull request merged**